lib/hooks.js

```diff
@@ -171,67 +172,124 @@ function hookRunnerApplication (hookName, boot, server, cb) {
   }
 }
 
-function hookRunner (functions, runner, request, reply, cb) {
-  let i = 0
+function hookRunnerGenerator (iterator) {
+  return function hookRunner (functions, request, reply, cb) {
+    let i = 0
+
+    function next (err) {
+      if (err || i === functions.length) {
+        cb(err, request, reply)
+        return
+      }
+
+      let result
+      try {
+        result = iterator(functions[i++], request, reply, next)
+      } catch (error) {
+        cb(error, request, reply)
+        return
+      }
+      if (result && typeof result.then === 'function') {
+        result.then(handleResolve, handleReject)
+      }
+    }
+
+    function handleResolve () {
+      next()
+    }
+
+    function handleReject (err) {
+      if (!err) {
+        err = new FST_ERR_SEND_UNDEFINED_ERR()
+      }
 
-  function next (err) {
-    if (err || i === functions.length) {
       cb(err, request, reply)
+    }
+
+    next()
+  }
+}
+
+function onResponseHookIterator (fn, request, reply, next) {
+  return fn(request, reply, next)
+}
+
+const onResponseHookRunner = hookRunnerGenerator(onResponseHookIterator)
+const onPreValidationHookRunner = hookRunnerGenerator(hookIterator)
+const onPreHandlerHookRunner = hookRunnerGenerator(hookIterator)
+const onTimeoutHookRunner = hookRunnerGenerator(hookIterator)
+const onRequestHookRunner = hookRunnerGenerator(hookIterator)
+
+function onSendHookRunner (functions, request, reply, payload, cb) {
+  let i = 0
+
+  function next (err, newPayload) {
+    if (err) {
+      cb(err, request, reply, payload)
+      return
+    }
+
+    if (newPayload !== undefined) {
+      payload = newPayload
+    }
+
+    if (i === functions.length) {
+      cb(null, request, reply, payload)
       return
     }
 
     let result
     try {
-      result = runner(functions[i++], request, reply, next)
+      result = functions[i++](request, reply, payload, next)
     } catch (error) {
-      next(error)
+      cb(error, request, reply)
       return
     }
     if (result && typeof result.then === 'function') {
       result.then(handleResolve, handleReject)
     }
   }
 
-  function handleResolve () {
-    next()
+  function handleResolve (newPayload) {
+    next(null, newPayload)
   }
 
   function handleReject (err) {
     if (!err) {
       err = new FST_ERR_SEND_UNDEFINED_ERR()
     }
 
-    cb(err, request, reply)
+    cb(err, request, reply, payload)
   }
 
   next()
 }
 
-function onSendHookRunner (functions, request, reply, payload, cb) {
+function preParsingHookRunner (functions, request, reply, cb) {
   let i = 0
 
   function next (err, newPayload) {
-    if (err) {
-      cb(err, request, reply, payload)
+    if (reply.sent) {
       return
     }
 
-    if (newPayload !== undefined) {
-      payload = newPayload
+    if (typeof newPayload !== 'undefined') {
```

Comment:
preParsing differs here from onSend. While onSend aborts before assigning the payload, preParsing is aborting after assigning the payload.

I assume, that we dont have any effects of this difference, because when errors happen the whole hook chain gets stopped.